### DISCUSSION TITLE
Read needs a CR/LF

### DISCUSSION THREAD

**Discussion by RaulHuertas** [serialport/serialport-rs#80] [Q&A] [answered]
Hello! I need to implment a binary protocol but my serialport only returns data if the other extreme sends a CR/LF sequence. How can I solve this?
thx!

```


use std::io::{self, Write};
use std::time::Duration;
use clap::Parser;
#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
	/// Name of the person to greet
	#[arg(short, long, default_value_t = String::from("/dev/ttyS4"))]
	port: String,

	/// Number of times to greet
	#[arg(short, long, default_value_t = 115200)]
	baudrate: u32,
}

fn main() {
	let args = Args::parse();
	let puerto = args.port;

	let puerto_serial = serialport::new(puerto.clone(), args.baudrate)
        		.timeout(Duration::from_millis(3))
        		.open();

	
	match puerto_serial {
		Ok(mut port) => {
			let mut serial_buf: Vec<u8> = vec![0; 1];
			println!("Receiving data on {} at {} baud:", &puerto, &args.baudrate);
			
			loop {
				match port.read(serial_buf.as_mut_slice()) {
					Ok(bytes_readed) => {
						io::stdout().write_all(&serial_buf[..bytes_readed]).expect("Log a stdout falló");
					}
					Err(ref e) => {
						if e.kind() == io::ErrorKind::TimedOut {
							//println!("timeout!");
						}
					}
				}
			}
		}
		Err(e) => {
			eprintln!("Failed to open {}. Error: {}", puerto, e);
			::std::process::exit(1);
		}
	}

}

```

**Comment by mlsvrts** [comment]
The operating system should buffer incoming serial data once the port has been opened. If I understand your question correctly, you should simply be able to write CR\LF to the port before reading the response.

**Reply by RaulHuertas** [reply]
I'm writing the reader. The writer is a device I can't modify :(

**Reply by mlsvrts** [reply]
Maybe I don't understand your problem? If you can't trigger data transmission, then there is nothing to do but wait until data arrives.

**Reply by RaulHuertas** [reply]
I'm trying that rust code with other tools, like Cutecom and screen. The rust code opens ttyS4, and these tools are tested in ttyS5. I have a cable connecting them.
Scenarios:
Cutecom(S5) -> (rust, S4) : Data is readed ONLY when I add a CR/LF termination to the lines I send. When I send multiple lines(A,B,C...) nothing is readed. If then I send a line X with CR/LF termination, the rust program receives ABC..X just after sending X.
Cutecom(S5) -> (screen, S4): screen always shows the dat inmediatly, with or without CR/LF

I think there is some buffering happening, but i don't know how to disable. It acts as if  was doing a "read_line" operation, but i am not.

**Reply by RaulHuertas** [reply]
I use those tools for testing. The real device i'll use doesnt send CR/LF

**Reply by mlsvrts** [reply]
I think I understand now. Can you please call `io::stdout().flush()` after your write and see if the behavior changes?

**Reply by RaulHuertas** [reply]
I don't write anything in rust :(

**Reply by mlsvrts** [reply]
I mean that instead of:

```rust
io::stdout().write_all(&serial_buf[..bytes_readed]).expect("Log a stdout falló");
```

You should try:

```rust
io::stdout().write_all(&serial_buf[..bytes_readed]).expect("Log a stdout falló");
io::stdout().flush();
```

**Reply by RaulHuertas** [reply]
I Know. but I'm not invoking write in my code. Just read()

**Reply by RaulHuertas** [reply, marked as answer]
OK. I solved it. Although i'm not writin anything to my serial port, I am writing to stdout. Since the binary data i'm receiving are not full lines, stdout doesn't flush.

So I DO receive data through the serial port, I'm just not flusing stdout. Thanks for your help!
